finality-aleph/src/sync/mod.rs

```diff
@@ -0,0 +1,92 @@
+use std::{
+    fmt::{Debug, Display},
+    hash::Hash,
+};
+
+mod substrate;
+
+/// The identifier of a block, the least amount of knowledge we can have about a block.
+pub trait BlockIdentifier: Clone + Hash + Debug + Eq {
+    /// The block number, useful when reasoning about hopeless forks.
+    fn number(&self) -> u32;
+}
+
+/// Informs the sync that it should attempt to acquire the specified data.
+pub trait Requester<BI: BlockIdentifier> {
+    /// The sync should attempt to acquire justifications for this block.
+    fn request_justification(&self, id: BI);
+}
+
+/// The header of a block, containing information about the parent relation.
+pub trait Header: Clone {
+    type Identifier: BlockIdentifier;
+
+    /// The identifier of this block.
+    fn id(&self) -> Self::Identifier;
+
+    /// The identifier of this block's parent.
+    fn parent_id(&self) -> Option<Self::Identifier>;
+}
+
+/// The justification of a block, including a header.
+pub trait Justification: Clone {
+    type Header: Header;
+
+    /// The header of the block.
+    fn header(&self) -> &Self::Header;
+}
+
+/// A verifier of justifications.
+pub trait Verifier<RJ, J: Justification> {
+    type Error: Display;
+
+    /// Verifies the raw justification and returns a full justification if successful, otherwise an
+    /// error.
+    fn verify(&self, header: J::Header, raw_justification: RJ) -> Result<J, Self::Error>;
+}
+
+/// A facility for finalizing blocks using justifications.
+pub trait Finalizer<J: Justification> {
+    type Error: Display;
+
+    /// Finalize a block using this justification. Since the justification contains the header, we
+    /// don't need to additionally specify the block.
+    fn finalize(&self, justification: J) -> Result<(), Self::Error>;
```

Comment:
suggestion, use anyhow + optionally thiserror

Reply:
I would rather not add these, if we want them either we should get an ADR or have people who care more add them personally. :P

Reply:
xD